### PR TITLE
feat: upload provenance to release via a `tag-name`

### PR DESCRIPTION
**Pull request by laurentsimon**
closes https://github.com/slsa-framework/slsa-github-generator/issues/710

### CODE REVIEW

File: .github/workflows/generator_generic_slsa3.yml
Comment (asraa) [inline .github/workflows/generator_generic_slsa3.yml:161]: I'm a little worried people will assume they need `upload-assets` to be true when they specify an `upload-url` 

Maybe we can clarify that in the description? Or allow that case, and only run this job when `upload-url` is nil.

Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml:161]: I agree it's confusing. How about requiring `upload-assets` be true *always* to an upload to occur? 

If `upload-assets`, the upload will occur automatically on new tags. If not new tag, users need this optional `upload-url` to indicate where to upload. 

Wdut?

Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml:166]: I've updated the doc to show what it would look like for the suggestion I made.

File: .github/workflows/builder_go_slsa3.yml
Comment (laurentsimon) [inline .github/workflows/builder_go_slsa3.yml:40]: @olivekl any feedback on the change in the file?
We're adding a new option called `upload-url` in this PR.
`upload-assets` is the option we already had, and it indicates that the workflow should automatically upload the provenance / binaries on new tags / releases.

`upload-url` is a new option for users to pass an upload url, because they sometimes create the release themselves in their workflow. In this case the workflow does not run on new release, it runs on a trgger event like `workflow_dispatch` (manual trigger) and the workflow creates the release before handing over the execution to our sSLSA generator.

/cc @asraa

File: .github/workflows/builder_go_slsa3.yml
Comment (asraa) [inline .github/workflows/builder_go_slsa3.yml:327]: check if `upload-assets` is true?

File: .github/workflows/builder_go_slsa3.yml
Comment (asraa) [inline .github/workflows/builder_go_slsa3.yml:294]: what if this is run with `upload-url` specified too? Should we double check that it is not?

**Review by asraa — COMMENTED** [review]
Thanks! I like this change. Let me know about the failsafing comment

File: .github/workflows/builder_go_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/builder_go_slsa3.yml]: done.

File: .github/workflows/builder_go_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/builder_go_slsa3.yml:294]: I've left it as-is to keep it simple.
If `upload-assets: true`, always upload for tags
if `upload-url` is also set, upload to the URL. This way is a user wants to upload to a different url from the tag, they can. And if they pass both, we will upload twice... which still works.

Wdut?

File: .github/workflows/generator_generic_slsa3.yml
Comment (ianlewis) [inline .github/workflows/generator_generic_slsa3.yml:171]: So when `upload-assets=true` and `upload-url` is set we upload twice? once as we do now (if this is a `workflow_dispatch` I guess this is a noop?), and once at the upload url specified?

File: .github/actions/upload-assets/package.json
Comment (ianlewis) [inline .github/actions/upload-assets/package.json:2]: upload-assets? or maybe rename the action's directory to `upload-release-asset`?

File: .github/workflows/generator_generic_slsa3.yml
Comment (ianlewis) [inline .github/workflows/generator_generic_slsa3.yml:191]: What was the reason we needed to create our own action for this?

File: .github/workflows/generator_generic_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml:171]: you''re correct. I've tried to not add much logic here, since the case you describe will hardly ever happen... I argue :)

But I can update if you think it's important.

Just let me know.

File: .github/actions/upload-assets/package.json
Comment (laurentsimon) [reply .github/actions/upload-assets/package.json:2]: will do thanks

File: .github/workflows/generator_generic_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml:191]: There is no available and maintained action to do that. It seems to not be a common case. The action that job project uses is https://github.com/actions/upload-release-asset, but it's not maintained. I've just copied their code and updated to latest dependency versions.

File: .github/workflows/generator_generic_slsa3.yml
Comment (ianlewis) [reply .github/workflows/generator_generic_slsa3.yml:171]: Based on your other comments it seemed like `upload-url` would only have an effect if `upload-assets=true` so I'm a bit confused.

Just to be clear, we expect them to set one or the other, but not both?

File: .github/workflows/generator_generic_slsa3.yml
Comment (ianlewis) [reply .github/workflows/generator_generic_slsa3.yml:191]: Since the normal supported wa to upload to releases is to use the `softprops/action-gh-release` and that action has a `tag_name` in order to support this use case, I wonder if we should just accept a `tag-name` input instead of `upload-url` and pass it along to `softprops/action-gh-release` (that we already call)?

Comment (ianlewis) [reply .github/workflows/generator_generic_slsa3.yml:191]: I fear we are adding a new action that we need to support just because our user jib is using the `upload-url` option on the unsupported action. I'm sure jib could update to use a `tag-name` to specify the release on `workflow_dispatch` events.

File: .github/workflows/generator_generic_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml:171]: so we expect `upload-assets=true` to always be set, optionally with a url

File: .github/workflows/generator_generic_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml:191]: Great idea. I'll re-work this PR then. Thanks for the suggestion

Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml]: Changes made. PTAL

File: .github/workflows/generator_generic_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/generator_generic_slsa3.yml]: no longer relevant as per https://github.com/slsa-framework/slsa-github-generator/pull/713#discussion_r960293407

File: .github/workflows/builder_go_slsa3.yml
Comment (ianlewis) [inline .github/workflows/builder_go_slsa3.yml:43]: Update this since we don't have an `upload-url` argument anymore.

File: .github/workflows/generator_generic_slsa3.yml
Comment (ianlewis) [inline .github/workflows/generator_generic_slsa3.yml:212]:
```suggestion
        id: release
```

**Review by ianlewis — APPROVED** [review]

**Comment by laurentsimon** [comment]
@ianlewis yamlling seems to be failing. Does
```
Successfully installed pathspec-0.10.1 yamllint-1.26.3
Error: [trailing-spaces] trailing spaces
```

means it found a problem with a yaml file? Or is it failing itself?

**Comment by ianlewis** [comment]
> @ianlewis yamlling seems to be failing. Does
> 
> ```
> Successfully installed pathspec-0.10.1 yamllint-1.26.3
> Error: [trailing-spaces] trailing spaces
> ```
> 
> means it found a problem with a yaml file? Or is it failing itself?

The error shows up in the PR diff view. I don't think the output shows up in the log very well for some reason. It seems to omit the file name and line number.
https://github.com/slsa-framework/slsa-github-generator/pull/713/files#diff-a526f8e747aef206fec2ce8469bbc904ee1a8d047bbb486743cca4e2dd5e5037R175

File: .github/workflows/generator_generic_slsa3.yml
Comment (ianlewis) [inline .github/workflows/generator_generic_slsa3.yml:175]: Fix trailing spaces
```suggestion
  # upload-assets uploads provenance to the release
```

**Review by ianlewis — APPROVED** [review]

File: .github/workflows/pre-submit.actions.yml
Comment (ianlewis) [inline .github/workflows/pre-submit.actions.yml:12]: If the name of a job changes then the protected branch's required pre-submits need to change as well. Ideally GitHub would allow you to set required pre-submits by an ID rather than name but it seems to use the name. We should think of job names as IDs.

I'm assuming that you're changing the name here just to add capitalization for consistency's sake? It's kind of hard keeping track of which of these are required for pre-submits so I suggest reverting the names and adding a comment that they shouldn't be changed unless there is a really good reason.

File: .github/workflows/pre-submit.units.yml
Comment (ianlewis) [inline .github/workflows/pre-submit.units.yml:35]: See other comment about required pre-submits

**Comment by asraa** [comment]
ping -- does this need another review after a merge?

**Comment by laurentsimon** [comment]
Let me get back to this PR when I'm back. It's been a while so I need to look into it. Thanks for pinging!

**Pull request merged**
